src/screencast/wlr_screencast.c

```diff
@@ -545,28 +546,146 @@ static void wlr_remove_output(struct xdpw_wlr_output *out) {
+	struct xdpw_screencast_context *ctx = data;
+
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_format_table called");
+
+	wlr_format_modifier_pair_emtpy_list(ctx);
+
+	ctx->feedback_data.format_table_data = mmap(NULL , size, PROT_READ, MAP_PRIVATE, fd, 0);
+	if (ctx->feedback_data.format_table_data == MAP_FAILED) {
+		ctx->feedback_data.format_table_data = NULL;
+		ctx->feedback_data.format_table_size = 0;
+		return;
+	}
+	ctx->feedback_data.format_table_size = size;
+}
+
+static void linux_dmabuf_feedback_handle_done(void *data,
+		struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1) {
+	struct xdpw_screencast_context *ctx = data;
+
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_handle_done called");
+	munmap(ctx->feedback_data.format_table_data, ctx->feedback_data.format_table_size);
+	ctx->feedback_data.format_table_data = NULL;
+	ctx->feedback_data.format_table_size = 0;
+}
+
+static void linux_dmabuf_feedback_tranche_done(void *data,
+		struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1) {
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_tranche_done called");
+}
+
+static void linux_dmabuf_feedback_tranche_formats(void *data,
+		struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1, struct wl_array *indices) {
+	struct xdpw_screencast_context *ctx = data;
+
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_format_table called");
+
+	if (!ctx->feedback_data.used_device) {
+		return;
+	}
+	struct fm_entry {
+		uint32_t format;
+		uint32_t padding;
+		uint64_t modifier;
+	};
+	// An entry in the table has to be 16 bytes long
+	assert(sizeof(struct fm_entry) == 16);
+
+	uint32_t n_modifiers = ctx->feedback_data.format_table_size/sizeof(struct fm_entry);
+	struct fm_entry *fm_entry = ctx->feedback_data.format_table_data;
+	uint16_t *idx;
+	wl_array_for_each(idx, indices) {
+		if (*idx >= n_modifiers) {
+			continue;
+		}
+		wlr_format_modifier_pair_add(ctx, (fm_entry + *idx)->format, (fm_entry + *idx)->modifier);
+	}
+}
+
+static void linux_dmabuf_feedback_tranche_flags(void *data,
+		struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1, uint32_t flags) {
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_tranche_flags called");
+}
+
+static void linux_dmabuf_feedback_tranche_target_devices(void *data,
+		struct zwp_linux_dmabuf_feedback_v1 *zwp_linux_dmabuf_feedback_v1, struct wl_array *device_arr) {
+	struct xdpw_screencast_context *ctx = data;
+
+	logprint(DEBUG, "wlroots: linux_dmabuf_feedback_tranche_target_devices called");
+	dev_t device;
+	assert(device_arr->size == sizeof(device));
+	memcpy(&device, device_arr->data, sizeof(device));
+
+	drmDevice *drmDev;
+	if (drmGetDeviceFromDevId(device, /* flags */ 0, &drmDev) < 0) {
+		return;
+	}
+	drmDevice *drmDevRenderer = NULL;
+	drmGetDevice2(gbm_device_get_fd(ctx->gbm), /* flags */ 0, &drmDevRenderer);
+	ctx->feedback_data.used_device = drmDevicesEqual(drmDevRenderer, drmDev) > 0;
```

Comment:
@emersion  Is this the right way to compare them?

Reply:
The return of [drmDevicesEqual](https://gitlab.freedesktop.org/mesa/drm/-/blob/main/xf86drm.c#L3719) are compatible with a bool, so we can just assign the return value directly.